apps/emqx_authn/src/emqx_authn_initializer.erl

```diff
@@ -0,0 +1,141 @@
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_authn_initializer).
+
+-behaviour(gen_statem).
+
+-define(DEFAULT_INTERVAL, 5).
+
+-include("emqx_authn.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%% APIs
+-export([start_link/0,
+         start_link/1,
+         initialize/0,
+         deinitialize/0
+        ]).
+
+%% gen_statem callbacks
+-export([init/1,
+         callback_mode/0
+        ]).
+
+%% state functions
+-export([waiting_for_process/3,
+         initializing/3,
+         idle/3
+        ]).
+
+-type(opts() :: #{name := atom(),
+                  interval := non_neg_integer(),
+                  init_fun := fun((pid()) -> any())}).
+
+%%--------------------------------------------------------------------
+%% External interface
+%%--------------------------------------------------------------------
+
+-spec(start_link() -> gen_statem:start_ret()).
+start_link() ->
+        start_link(
+          #{name => ?AUTHN,
+            interval => ?DEFAULT_INTERVAL,
+            init_fun => fun(_) -> initialize() end}).
+
+-spec(start_link(opts()) -> gen_statem:start_ret()).
+start_link(#{} = Opts) ->
+        gen_statem:start_link(?MODULE, Opts, []).
+
+-spec(initialize() -> ok).
+initialize() ->
+    _ = ?AUTHN:register_providers(emqx_authn:providers()),
+
+    lists:foreach(
+      fun({ChainName, RawAuthConfigs}) ->
+              AuthConfig = emqx_authn:check_configs(RawAuthConfigs),
+              ?AUTHN:initialize_authentication(
+                 ChainName,
+                 AuthConfig)
+      end,
+      chain_configs()).
+
+-spec(deinitialize() -> ok).
+deinitialize() ->
+    ok = ?AUTHN:deregister_providers(provider_types()).
+
+%%--------------------------------------------------------------------
+%% gen_statem callbacks
+%%--------------------------------------------------------------------
+
+callback_mode() -> state_functions.
+
+init(#{init_fun := Fun,
+       name := Name,
+       interval := Interval} = Data)
+  when
+      is_function(Fun, 1)
+      and is_atom(Name)
+      and is_integer(Interval) ->
+    {ok, waiting_for_process, Data, [{state_timeout, 0, check_process}]}.
+
+waiting_for_process(state_timeout, check_process, #{name := Name,
+                                                   interval := Interval} = Data) ->
+    case whereis(Name) of
+        undefined ->
+            ?tp(info, emqx_authn_initializer_no_process, #{name => Name}),
```

Comment:
With the default interval of [5](https://github.com/emqx/emqx/blob/2cb9955251fec03c0f3ab48b96a90c34a0bf2c66/apps/emqx_authn/src/emqx_authn_initializer.erl#L21) ms, could this generate a large volume of logs?

Reply:
✅ Lowered severity to debug here, it's unlikely that retries are important for info logs.